proto/src/protobuf.rs

```diff
@@ -1,8 +1,20 @@
 /// A marker trait that captures the relationships between a domain type (`Self`) and a protobuf type (`P`).
 pub trait Protobuf<P>: Sized
 where
-    P: prost::Message,
+    P: prost::Message + Default,
     P: std::convert::From<Self>,
-    Self: std::convert::TryFrom<P>,
+    Self: std::convert::TryFrom<P> + Clone,
+    <Self as std::convert::TryFrom<P>>::Error: Into<anyhow::Error>,
 {
+    /// Encode this domain type to a byte vector, via proto type `P`.
+    fn encode_to_vec(&self) -> Vec<u8> {
+        P::from(self.clone()).encode_to_vec()
+    }
+
+    /// Decode this domain type from a byte buffer, via proto type `P`.
+    fn decode<B: bytes::Buf>(buf: B) -> Result<Self, anyhow::Error> {
+        <P as prost::Message>::decode(buf)?
+            .try_into()
+            .map_err(Into::into)
+    }
```

Comment:
This means that we don't have to manually implement a ton of `Vec<u8>` conversions for each domain type; I think there's some like that on `Transaction` that could now be cleaned up.